linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -698,45 +695,97 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
       tcaptureValues: List[PreTransform])(cont: PreTransCont)(
       implicit scope: Scope, pos: Position): TailRec[Tree] = {
 
-    val thisType = if (arrow) NoType else AnyType
-
-    val (allNewParamsWithUse, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body)
-
-    val (newCaptureParamsWithUse, newParamsWithUse) = allNewParamsWithUse.splitAt(captureParams.size)
-
-    val (newParams, newRestParam) = {
-      val t = newParamsWithUse.map(_._1).toList
-      if (restParam.isDefined) (t.init, Some(t.last))
-      else (t, None)
+    val (paramLocalDefs, newParams) = params.map(newParamReplacement(_)).unzip
+    val (restParamLocalDef, newRestParam) = {
+      // Option#unzip
+      restParam.map(newParamReplacement(_)) match {
+        case None         => (None, None)
+        case Some((x, y)) => (Some(x), Some(y))
+      }
     }
 
-    withCaptures(newCaptureParamsWithUse, tcaptureValues) { (finalCaptureParams, finalCaptureValues, cont1) =>
-      val newClosure = Closure(arrow, finalCaptureParams, newParams, newRestParam, newBody, finalCaptureValues)
-      val newTree = PreTransTree(newClosure, RefinedType(AnyType, isExact = false, isNullable = false))
-      cont1(newTree)
+    val thisLocalDef =
+      if (arrow) None
+      else Some(newThisReplacement(AnyType))
+
+    val innerEnv = OptEnv.Empty
+      .withThisLocalDef(thisLocalDef)
+      .withLocalDefs(paramLocalDefs)
+      .withLocalDefs(restParamLocalDef.toList)
+
+    transformCapturingBody(captureParams, tcaptureValues, body, innerEnv) {
+        (newCaptureParams, newCaptureValues, newBody) =>
+      val newClosure = Closure(arrow, newCaptureParams, newParams, newRestParam, newBody, newCaptureValues)
+      PreTransTree(newClosure, RefinedType(AnyType, isExact = false, isNullable = false))
```

Comment:
Same indentation issue here:
```suggestion
      (newCaptureParams, newCaptureValues, newBody) =>
        val newClosure = Closure(arrow, newCaptureParams, newParams, newRestParam, newBody, newCaptureValues)
        PreTransTree(newClosure, RefinedType(AnyType, isExact = false, isNullable = false))
```